src/napi.h

```diff
@@ -0,0 +1,663 @@
+   * to traverse a C++ stack and be thrown and caught as a C++ exception.
+   *
+   * If a NAPI API call fails without executing any JavaScript code (for example due
+   * to an invalid argument), then the NAPI wrapper automatically converts and throws
+   * the error as a C++ exception of type Napi::Error.
+   *
+   * If a JavaScript function called by C++ code via NAPI throws a JavaScript exception,
+   * then the NAPI wrapper automatically converts and throws it as a C++ exception of type
+   * Napi::Error.
+   *
+   * If a C++ exception of type Napi::Error escapes from a NAPI C++ callback, then
+   * the NAPI wrapper automatically converts and throws it as a JavaScript exception.
+   *
+   * Catching a C++ exception of type Napi::Error also clears the JavaScript exception.
+   * Of course it may be then re-thrown, which restores the JavaScript exception.
+   *
+   * Example 1 - Throwing an exception:
+   *
+   *   Napi::Env env = ...
+   *   throw env.NewError("Example exception");
+   *   // Following C++ statements will not be executed.
+   *   // The exception will bubble up as a C++ exception of type Napi::Error,
+   *   // until it is either caught while still in C++, or else automatically
+   *   // re-thrown as a JavaScript exception when the callback returns to JavaScript.
+   *
+   * Example 2 - Ignoring a NAPI exception:
+   *
+   *   Napi::Function jsFunctionThatThrows = someObj.AsFunction();
+   *   jsFunctionThatThrows({ arg1, arg2 });
+   *   // Following C++ statements will not be executed.
+   *   // The exception will bubble up as a C++ exception of type Napi::Error,
+   *   // until it is either caught while still in C++, or else automatically
+   *   // re-thrown as a JavaScript exception when the callback returns to JavaScript.
+   *
+   * Example 3 - Handling a NAPI exception:
+   *
+   *   Napi::Function jsFunctionThatThrows = someObj.AsFunction();
+   *   try {
+   *      jsFunctionThatThrows({ arg1, arg2 });
+   *   }
+   *   catch (const Napi::Error& e) {
+   *     cerr << "Caught JavaScript exception: " + e.what();
+   *     // Since the exception was caught here, it will not be re-thrown as
+   *     // a JavaScript exception.
+   *   }
+   */
+  class Error : public Value, public std::exception {
+  public:
+    static Error New(Napi::Env env);
+    static Error New(Napi::Env env, const char* message);
+    static Error NewTypeError(Napi::Env env, const char* message);
+    static Error NewRangeError(Napi::Env env, const char* message);
+
+    Error();
+    Error(napi_env env, napi_value value);
+
+    std::string Message() const;
+    void ThrowAsJavaScriptException() const;
+
+    const char* what() const override;
+
+  private:
+    std::string _message;
+  };
+
+  class Persistent {
+  public:
+    Persistent();
+    Persistent(napi_env env, napi_persistent persistent);
+    ~Persistent();
+
+    // A persistent instance can be moved but cannot be copied.
```

Comment:
Unless I am missing some important detail, that sounds very dangerous. The reason V8 redid `Persistent`s in the version around Node 0.11 was to avoid the problem of careless leaks. The old `Persistent` API found in Node 0.10 allowed implicit copying of `Persistent` handles which led to a whole bunch of leaks due to misuse. It should be impossible to accidentally copy a `Persistent` handle.